core/dbt/cli/artifact_upload.py

```diff
@@ -0,0 +1,162 @@
+    "run": EXECUTION_ARTIFACTS,
+    "run-operation": EXECUTION_ARTIFACTS,
+    "seed": EXECUTION_ARTIFACTS,
+    "snapshot": EXECUTION_ARTIFACTS,
+    "test": EXECUTION_ARTIFACTS,
+    "freshness": [MANIFEST_FILE_NAME, SOURCE_RESULT_FILE_NAME],
+    "generate": [MANIFEST_FILE_NAME, CATALOG_FILENAME],
+}
+
+
+class ArtifactUploadConfig(BaseSettings):
+    tenant_hostname: str
+    DBT_CLOUD_TOKEN: str
+    DBT_CLOUD_ACCOUNT_ID: str
+    DBT_CLOUD_ENVIRONMENT_ID: str
+
+    def get_ingest_url(self):
+        return f"https://{self.tenant_hostname}/api/private/accounts/{self.DBT_CLOUD_ACCOUNT_ID}/environments/{self.DBT_CLOUD_ENVIRONMENT_ID}/ingests/"
+
+    def get_complete_url(self, ingest_id):
+        return f"{self.get_ingest_url()}{ingest_id}/complete/"
+
+    def get_headers(self, invocation_id=None):
+        if invocation_id is None:
+            invocation_id = str(uuid.uuid4())
+        return {
+            "Accept": "application/json",
+            "X-Invocation-Id": invocation_id,
+            "Authorization": f"Token {self.DBT_CLOUD_TOKEN}",
+        }
+
+
+def _retry_with_backoff(operation_name, func, max_retries=MAX_RETRIES, retry_codes=None):
+    """Execute a function with exponential backoff retry logic.
+
+    Args:
+        operation_name: Name of the operation for error messages
+        func: Function to execute that returns (success, result)
+        max_retries: Maximum number of retry attempts
+
+    Returns:
+        The result from the function if successful
+
+    Raises:
+        DbtException: If all retry attempts fail
+    """
+    if retry_codes is None:
+        retry_codes = [500, 502, 503, 504]
+    retry_delay = 1
+    for attempt in range(max_retries):
+        try:
+            success, result = func()
+            if success:
+                return result
+
+            if result.status_code not in retry_codes:
+                raise DbtException(f"Error {operation_name}: {result}")
+            if attempt == max_retries - 1:  # Last attempt
+                raise DbtException(f"Error {operation_name}: {result}")
+
+            time.sleep(retry_delay)
+            retry_delay *= 2  # exponential backoff
+        except requests.RequestException as e:
+            if attempt == max_retries - 1:  # Last attempt
+                raise DbtException(f"Error {operation_name}: {str(e)}")
+
+            time.sleep(retry_delay)
+            retry_delay *= 2  # exponential backoff
+
+
+def upload_artifacts(project_dir, target_path, command):
+    # Check if there are artifacts to upload for this command
+    if command not in ARTIFACTS_TO_UPLOAD:
+        fire_event(ArtifactUploadSkipped(msg=f"No artifacts to upload for command {command}"))
+        return
+
+    # read configurations
+    try:
+        project = load_project(
+            project_dir, version_check=False, profile=UnsetProfile(), cli_vars=None
+        )
+        if not project.dbt_cloud or "tenant_hostname" not in project.dbt_cloud:
```

Comment:
why do we audit for tenant_hostname but none of the expected env vars?

Reply:
The audit of env vars is handled at initialization of BaseSettings.